### PR TITLE
[AArch64][NEON][SVE] Lower i8 to i64 partial reduction to a dot product

### PR DESCRIPTION
**Pull request by JamesChesterman**
An i8 to i64 partial reduction can instead be done with an i8 to i32 dot product followed by a sign extension.

### CODE REVIEW

**Comment by JamesChesterman** [comment]
@huntergr-arm 
@SamTebbs33 
@MacDue 
@sdesmalen-arm 
@kmclaughlin-arm 
@NickGuy-Arm

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -21942,6 +21944,20 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
...
+  if ((ReducedType == MVT::nxv4i64 && MulSrcType == MVT::nxv16i8) ||
+      (ReducedType == MVT::v4i64 && MulSrcType == MVT::v16i8)) {
+    EVT ReducedTypeHalved = (ReducedType.isScalableVector()) ? MVT::nxv4i32 : MVT::v4i32;
+
+    auto Doti32 =
+        DAG.getNode(Opcode, DL, ReducedTypeHalved,
+                    DAG.getConstant(0, DL, ReducedTypeHalved), A, B);
+    auto Extended = DAG.getSExtOrTrunc(Doti32, DL, ReducedType);
+    return DAG.getNode(ISD::ADD, DL, NarrowOp.getValueType(),
+                                  {NarrowOp, Extended});
```
Comment (SamTebbs33) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21958]: I think we can make this slightly cleaner by omitting the brackets around the `NarrowOp` and `Extended` operands. There should be a variant of `getNode` that takes the two operands without them being in a list.

Comment (JamesChesterman) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21958]: Done

**Review by SamTebbs33 — APPROVED** [review]
Looks good to me!

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 2]
```diff
@@ -21942,6 +21944,20 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
   else
     Opcode = AArch64ISD::UDOT;
 
+  // Partial reduction lowering for (nx)v16i8 to (nx)v4i64 requires an i32 dot
+  // product followed by a zero / sign extension
+  if ((ReducedType == MVT::nxv4i64 && MulSrcType == MVT::nxv16i8) ||
+      (ReducedType == MVT::v4i64 && MulSrcType == MVT::v16i8)) {
+    EVT ReducedTypeHalved =
+        (ReducedType.isScalableVector()) ? MVT::nxv4i32 : MVT::v4i32;
```
Comment (MacDue) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21952]: (optional) nit:  `ReducedTypeHalved` -> `ReducedTypeI32`? 
To me `Halved` could mean half the vector width (e.g. v4 -> v2).

Comment (JamesChesterman) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Yes I agree, thank you for the advice, done.

**Review by MacDue — APPROVED** [review]
LGTM, thanks for the fixup :+1: 
P.s. You may want to rebase (since I think CI is failing due to unrelated changes)

**Pull request merged**